src/tools/rbd_mirror/image_map/Policy.h

```diff
@@ -0,0 +1,251 @@
+  void init(Context *on_finish);
+  // shut down map operations
+  void shut_down(Context *on_finish);
+
+  // update (add/remove) images
+  void update_images(const std::string &mirror_uuid,
+                     std::set<std::string> &&added_global_image_ids,
+                     std::set<std::string> &&removed_global_image_ids);
+
+  // handle notify response from remote peer (r: 0 == success, -ve otherwise)
+  void handle_peer_ack(const std::string &global_image_id, int r);
+
+  // add/remove instances
+  void update_instances(const std::vector<std::string> &instances);
+
+private:
+  // Lock ordering: m_threads->timer_lock, m_map_lock, m_lock
+
+  struct Update {
+    std::string global_image_id;
+    std::string instance_id;
+
+    Update(const std::string &global_image_id, const std::string &instance_id)
+      : global_image_id(global_image_id),
+        instance_id(instance_id) {
+    }
+  };
+  typedef std::list<Update> Updates;
+
+  struct Info {
+    utime_t last_shuffled_time;               // timestamp when this image
+                                              // was last shuffled
+    std::set<std::string> registered_peers;   // registered peers for this
+                                              // image ("" == local, otherwise)
+    std::set<std::string> unregistered_peers; // unregistered peers (remotes only)
+
+    std::string action_instance_id;           // instance id of next (sub)action
+    std::set<std::string> action_peers;       // set of peers (if any) for the
+                                              // next (sub)action
+
+    // set last_shuffled_time to zero so that as peer daemons are
+    // spawned we shuffle freshly mapped images.
+    Info()
+      : last_shuffled_time(utime_t(0, 0)) {
+    }
+  };
+
+  librados::IoCtx &m_ioctx;
+  Threads<ImageCtxT> *m_threads;
+  Listener &m_listener;
+  Context *m_timer_task;
+
+  Mutex m_lock;
+  Cond m_cond;
+  bool m_shutting_down;
+  AsyncOpTracker m_async_op_tracker;
+
+  std::set<std::string> m_global_image_ids;
+
+  ImageAction m_image_action;
+  std::map<std::string, Info> m_image_info;
+
+  struct C_LoadMap : public Context {
+    Policy *policy;
+    Context *on_finish;
+
+    std::map<std::string, cls::rbd::MirrorImageMap> image_mapping;
+
+    C_LoadMap(Policy *policy, Context *on_finish)
+      : policy(policy),
+        on_finish(on_finish) {
+    }
+
+    void finish(int r) override {
+      if (r == 0) {
+        policy->load(image_mapping);
+      }
+
+      on_finish->complete(r);
+      policy->finish_async_op();
+    }
+  };
+
+  struct C_NotifyListener : public Context {
+    Policy *policy;
+
+    Updates to_map;
+    Updates to_unmap;
+    Updates to_purge;
+
+    C_NotifyListener(Policy *policy)
+      : policy(policy) {
+    }
+
+    void finish(int r) override {
+      if (r == 0) {
+        policy->notify_listener(to_map, to_unmap, to_purge);
+      } else {
+        // on failure, resume pending actions (if any) for this
```

Comment:
On failure I don't think it would be safe to proceed to the action. If you failed to notify an instance, try again until that instance is removed (marked dead) so that you know it's safe to proceed.